pkg/k8shandler/cluster.go

```diff
@@ -307,6 +377,112 @@ func (cState *ClusterState) amendPods(dpl *v1alpha1.Elasticsearch) error {
+		if node.UnderUpgrade == v1alpha1.UnderUpgradeFalse {
+			return &node
+		}
+	}
+	return nil
+}
+
+func (cState *ClusterState) beginUpgrade(dpl *v1alpha1.Elasticsearch, owner metav1.OwnerReference) (*v1alpha1.ElasticsearchNodeStatus, error) {
+	masterPod, err := getRunningMasterPod(dpl.Name, dpl.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	if err = disableShardAllocation(dpl, masterPod); err != nil {
+		return nil, err
+	}
+	if err = utils.PerformSyncedFlush(masterPod); err != nil {
+		return nil, err
+	}
+	node, err := cState.upgradeNode(dpl, owner)
+	if err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+func (cState *ClusterState) upgradeNode(dpl *v1alpha1.Elasticsearch, owner metav1.OwnerReference) (*v1alpha1.ElasticsearchNodeStatus, error) {
+	nodeForUpgrade := selectNodeForUpgrade(dpl)
+	if nodeForUpgrade == nil {
+		// upgrade requested, but there are no nodes for upgrade?
+		return nil, fmt.Errorf("No nodes for upgrade found")
+	}
+	// node managed by Deployment/StatefulSet
+	nodeName := nodeForUpgrade.DeploymentName
+	if nodeName == "" {
+		nodeName = nodeForUpgrade.StatefulSetName
+	}
+	for _, node := range cState.Nodes {
+		if node.Desired.DeployName == nodeName {
+			// found match between current and desired node state - let's upgrade the node
+			if err := node.Desired.CreateOrUpdateNode(owner); err != nil {
+				return nil, fmt.Errorf("Unable to create Elasticsearch node: %v", err)
+			}
+			retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				// get the latest Elasticsearch resource
+				if getErr := sdk.Get(dpl); getErr != nil {
+					return getErr
+				}
+				// dpl updated - order of ElasticsearchNodeStatus might have changed
+				// we can't rely on array indices, have go through the array again
+				for index, node := range dpl.Status.Nodes {
+					if node.PodName == nodeForUpgrade.PodName {
+						dpl.Status.Nodes[index].UnderUpgrade = v1alpha1.UnderUpgradeTrue
+						nodeForUpgrade = &dpl.Status.Nodes[index]
+						break
+					}
+				}
+				return sdk.Update(dpl)
+			})
+			return nodeForUpgrade, retryErr
+		}
+	}
+	return nil, fmt.Errorf("No nodes for upgrade found")
+}
+
+func disableShardAllocation(dpl *v1alpha1.Elasticsearch, masterPod *v1.Pod) error {
+	return setShardAllocation(dpl, masterPod, false)
+}
+func enableShardAllocation(dpl *v1alpha1.Elasticsearch, masterPod *v1.Pod) error {
+	return setShardAllocation(dpl, masterPod, true)
+}
+
+func setShardAllocation(dpl *v1alpha1.Elasticsearch, masterPod *v1.Pod, enabled bool) error {
+	if err := utils.SetShardAllocation(masterPod, enabled); err != nil {
+		return err
+	}
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		if getErr := sdk.Get(dpl); getErr != nil {
+			return getErr
+		}
+		if dpl.Status.ShardAllocationEnabled == enabled {
+			return nil
+		}
+		dpl.Status.ShardAllocationEnabled = enabled
+		return sdk.Update(dpl)
+	})
+	// TODO: should we revert shard allocation?
```

Comment:
i think there are other conditions (TBD) that would make us want to reenable during an upgrade like to prevent a node from falling over because of the influx of logs and recovery taking too long?